36 - Integración API Externa

paul-ot opened the pull request:
- Añadido endpoint para hacer GET de libros por su ISBN.
- Añadido nuevo controlador para get de libros desde Open Library.
- Añadida clase en app/services, para comunicarnos con Open Library.
- Añadida gema HTTParty, para facilitar el uso de http requests.

paul-ot committed:
- Añadido endpoint para hacer GET de libros por su ISBN.
- Añadido nuevo controlador para get de libros desde Open Library.
- Añadida clase en app/services, para comunicarnos con Open Library.
- Añadida gema HTTParty, para facilitar el uso de http requests.

holywyvern commented on app/services/open_library.rb:
```diff
@@ -0,0 +1,16 @@
+
+  def find_by_isbn(isbn)
+    @options[:query][:bibkeys] = "ISBN:#{isbn}"
+    response = self.class.get('/books', @options)
```
Que pasa cuando consulto por un ISBN que no existe ?

Deberías agregar un error del estilo:

`raise OpenLibraryError unless response.ok?`

y usar `rescue_from`

paul-ot replied:
Ah, bien, cuando no existía, la clase devolvía un objeto vacío `{}`, y controlaba  eso en el Controlador, ahora lo arreglo.

paul-ot replied:
El problema es que por más que no exista el ISBN, la API retorna siempre ok (http:200), que te parece si hago algo como `raise OpenLibraryError if response.body == '{}'`? @holywyvern

holywyvern replied:
Creo que es mejor que hagas algo como:

```
json = JSON.parse(response.body)
raise OpenLibraryError if json.blank?
json
```

vos @mtejedorwolox, como preferis manejarlo ?

Igual lo que planteaste esta "bien" pero me hace un poco de ruido comparar strings.

paul-ot replied:
Implementé como vos lo decís @holywyvern.

paul-ot committed:
- Added a new PORO (an exception class to deal with book not found exception).
- Added 'app/poros' directory to the autoload paths.
- Implemented rescue for OpenLibraryException.
- Implemented raise for OpenLibraryException in the service class services/open_library

paul-ot committed:
Merge branch 'development' into 36-integración-api-externa
